language-plutus-napkin/src/Language/PlutusCore/Type.hs

```diff
@@ -0,0 +1,153 @@
+{-# LANGUAGE DeriveAnyClass    #-}
+{-# LANGUAGE DeriveFoldable    #-}
+{-# LANGUAGE DeriveFunctor     #-}
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE KindSignatures    #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TypeFamilies      #-}
+
+module Language.PlutusCore.Type ( Term (..)
+                                , Type (..)
+                                , Token (..)
+                                , Builtin (..)
+                                , Kind (..)
+                                , Keyword (..)
+                                , Special (..)
+                                , Name (..)
+                                , Version (..)
+                                , Program (..)
+                                -- * Base functors
+                                , TermF (..)
+                                , TypeF (..)
+                                ) where
+
+import qualified Data.ByteString.Lazy           as BSL
+import           Data.Functor.Foldable          (cata)
+import           Data.Functor.Foldable.TH
+import           Data.Text.Encoding             (decodeUtf8)
+import           Data.Text.Prettyprint.Doc
+import           Language.PlutusCore.Identifier
+import           PlutusPrelude
+
+data Builtin = AddInteger
+             | SubtractInteger
+             | MultiplyInteger
+             | DivideInteger
+             | RemainderInteger
+             | LessThanInteger
+             | LessThanEqInteger
+             | GreaterThanInteger
+             | GreaterThanEqInteger
+             | EqInteger
+             | IntToByteString
+             | Ceiling
+             | Floor
+             | Round
+             | Concatenate
+             | TakeByteString
+             | DropByteString
+             | SHA2
+             | SHA3
+             | VerifySignature
+             | EqByteString
+             | TxHash
+             | BlockNum
+             | BlockTime
+             deriving (Show, Eq, Generic, NFData)
+
+data Version a = Version a Integer Integer Integer
+               deriving (Generic, NFData)
+
+data Keyword = KwIsa
+             | KwAbs
+             | KwInst
+             | KwLam
+             | KwFix
+             | KwBuiltin
+             | KwFun
+             | KwForall
+             | KwByteString
+             | KwInteger
+             | KwSize
+             | KwType
+             | KwProgram
+             deriving (Show, Eq, Generic, NFData)
+
+data Special = OpenParen
+             | CloseParen
+             | OpenBracket
+             | CloseBracket
+             | Dot
+             deriving (Show, Eq, Generic, NFData)
+
+-- | Annotated type for names
+data Token a = LexName { loc :: a, identifier :: Unique }
```

Comment:
I think, we should have two separate modules for (a) the types that are internal, like the various tokens etc, and (b) for the Plutus Core AST, which will be part of the external library API. (I think, it is also conceptually cleaner as these two data structures are used in different stages of the compilation pipeline.)